test/unit/gateways/firstdata_e4_v27_test.rb

```diff
@@ -0,0 +1,980 @@
+  <Bank_Message>Approved</Bank_Message>
+  <Bank_Resp_Code_2></Bank_Resp_Code_2>
+  <SequenceNo>000040</SequenceNo>
+  <AVS>1</AVS>
+  <CVV2>M</CVV2>
+  <Retrieval_Ref_No>7228838</Retrieval_Ref_No>
+  <CAVV_Response></CAVV_Response>
+  <Currency>USD</Currency>
+  <AmountRequested></AmountRequested>
+  <PartialRedemption>false</PartialRedemption>
+  <MerchantName>FriendlyInc</MerchantName>
+  <MerchantAddress>123 Main Street</MerchantAddress>
+  <MerchantCity>Durham</MerchantCity>
+  <MerchantProvince>North Carolina</MerchantProvince>
+  <MerchantCountry>United States</MerchantCountry>
+  <MerchantPostal>27592</MerchantPostal>
+  <MerchantURL></MerchantURL>
+  <TransarmorToken></TransarmorToken>
+  <CardType>Visa</CardType>
+  <CurrentBalance></CurrentBalance>
+  <PreviousBalance></PreviousBalance>
+  <EAN></EAN>
+  <CardCost></CardCost>
+  <VirtualCard>false</VirtualCard>
+  <CTR>=========== TRANSACTION RECORD ==========
+FriendlyInc DEMO0
+123 Main Street
+Durham, NC 27592
+United States
+
+
+TYPE: Auth Only
+
+ACCT: Visa  $ 0.00 USD
+
+CARDHOLDER NAME : Longbob Longsen
+CARD NUMBER     : ############4242
+DATE/TIME       : 04 Jul 14 14:21:52
+REFERENCE #     :  000040 M
+AUTHOR. #       : ET184931
+TRANS. REF.     : 1
+
+    Approved - Thank You 100
+
+
+Please retain this copy for your records.
+
+Cardholder will pay above amount to card
+issuer pursuant to cardholder agreement.
+=========================================</CTR>
+</TransactionResult>
+    RESPONSE
+  end
+
+  def no_transaction_response
+    yamlexcep = <<-RESPONSE
+--- !ruby/exception:ActiveMerchant::ResponseError
+message: Failed with 400 Bad Request
+message:
+response: !ruby/object:Net::HTTPBadRequest
+  body: "Malformed request: Transaction Type is missing."
+  body_exist: true
+  code: "400"
+  header:
+    connection:
+    - Close
+    content-type:
+    - text/html; charset=utf-8
+    server:
+    - Apache
+    date:
+    - Fri, 28 Sep 2012 18:21:37 GMT
+    content-length:
+    - "47"
+    status:
+    - "400"
+    cache-control:
+    - no-cache
+  http_version: "1.1"
+  message: Bad Request
+  read: true
+  socket:
+    RESPONSE
+    YAML.load(yamlexcep)
+  end
+
+  def bad_credentials_response
+    yamlexcep = <<-RESPONSE
+--- !ruby/exception:ActiveMerchant::ResponseError
+message:
+response: !ruby/object:Net::HTTPUnauthorized
+  code: '401'
+  message: Authorization Required
+  body: Unauthorized Request. Bad or missing credentials.
+  read: true
+  header:
+    cache-control:
+    - no-cache
+    content-type:
+    - text/html; charset=utf-8
+    date:
+    - Tue, 30 Dec 2014 23:28:32 GMT
+    server:
+    - Apache
+    status:
+    - '401'
+    x-rack-cache:
+    - invalidate, pass
+    x-request-id:
+    - 4157e21cc5620a95ead8d2025b55bdf4
+    x-ua-compatible:
+    - IE=Edge,chrome=1
+    content-length:
+    - '49'
+    connection:
+    - Close
+  body_exist: true
+  http_version: '1.1'
+  socket:
+    RESPONSE
+    YAML.load(yamlexcep)
```

Comment:
Rubocop suggests `YAML.safe_load()`. But using it throws an exception on objects referenced within the document.

> Psych::DisallowedClass: Tried to load unspecified class: Net::HTTPUnauthorized

Reply:
`YAML.load` will deserialize arbitrary Ruby classes if you specify the object with a `!` (so the `!ruby/exception:ActiveMerchant::ResponseError` here), which is really dangerous (imagine someone deserializing the `File` class with reading from `/etc/passwd` or the like). `safe_load` only allows whitelisted classes. In this case we control the YAML, so the RuboCop recommendation isn't _strictly_ necessary, but it's still a good habit. The only trick is you have to tell it which classes are okay. In this case, if you change `YAML.load(yamlexcep)` to `YAML.safe_load(yamlexcep, ['Net::HTTPBadRequest', 'ActiveMerchant::ResponseError'])` (based on the class serializations on line 857 and 859), both RuboCop and your code will be happy.